File: compute-invariants.pl
use strict;
use warnings;

use autodie;

use IO::Handle ();
use List::MoreUtils 'uniq';
use Math::Combinatorics 'combine';

my @invariants = get_invariants(g_getcells(g_getselrect()));
g_exit("@invariants");

sub get_invariants {
    my ($cells) = @_;

    my @invariants;
    foreach my $t (0 .. 3) {
        my @gliders;
        my $temp_cells = [ @$cells ];
        foreach my $gen (0 .. 3) {
            push @gliders, map { [ @$_, $gen ] } find_all_gliders($temp_cells);
            $temp_cells = g_evolve($temp_cells, 1);
        }

        my @glider_pairs = combine(2, @gliders);
        foreach my $pair (@glider_pairs) {
            push @invariants, compute_invariants(@$pair);
        }

        $cells = g_transform($cells, 0, 0, 0, -1, 1, 0);
    }

    return uniq(@invariants);
}

sub find_all_gliders {
    my ($cells) = @_;

    my $map = {};
    for (my $c = 0; $c < @$cells; $c += 2) {
        my ($x, $y) = @$cells[$c, $c + 1];
        $map->{$x}{$y}++;
    }

    my @ret;
    POSITION: for (my $c = 0; $c < @$cells; $c += 2) {
        my ($x, $y) = @$cells[$c, $c + 1];

        my $glider_map = {
            $x     => { $y     => 1, $y + 2 => 1 },
            $x + 1 => { $y + 1 => 1, $y + 2 => 1 },
            $x + 2 => { $y + 1 => 1 },
        };

        for (my $i = $x - 1; $i < $x + 4; $i++) {
            for (my $j = $y - 1; $j < $y + 4; $j++) {
                next POSITION if $map->{ $i }{ $j } xor $glider_map->{ $i }{ $j };
            }
        }

        push @ret, [$x, $y];
    }

    return @ret;
}

sub compute_invariants {
    my ($g1, $g2) = @_;

    my $dx = $g2->[0] - $g1->[0];
    my $dy = $g2->[1] - $g1->[1];
    my $dt = $g2->[2] - $g1->[2];

    my $hd = $dx - $dy;
    my $tr = $hd / 2 * 47;

    return abs($dt - 4 * $dy - $tr), abs(4 * $dx - $dt - $tr);
}

File: compute-invariants.py
import golly as g
from itertools import combinations, product

def get_invariants(cells):
  invariants = []

  for t in range(0, 4):
    gliders = []
    temp_cells = cells[:]

    for gen in range(0, 4):
      gliders.extend([coords + (gen,) for coords in find_all_gliders(temp_cells)])
      temp_cells = g.evolve(temp_cells, 1)

    glider_pairs = combinations(gliders, 2)
    invariants.extend([item for pair in glider_pairs for item in compute_invariants(pair)])

    cells = g.transform(cells, 0, 0, 0, -1, 1, 0)

  seen = set()
  return [ item for item in invariants if item not in seen and not seen.add(item) ]

def find_all_gliders(cells):
  gliders = []

  cells_dict = dict.fromkeys(zip(cells[::2], cells[1::2]))

  for x, y in zip(cells[::2], cells[1::2]):
    glider_dict = dict.fromkeys([(x, y), (x, y+2), (x+1, y+1), (x+1, y+2), (x+2, y+1)])

    spot = product(range(x-1, x+4), range(y-1, y+4))
    if all([ ((i, j) in cells_dict) is ((i, j) in glider_dict) for i, j in spot ]):
      gliders.append( (x, y) )

  return gliders

def compute_invariants(glider_pair):
  (x1, y1, t1), (x2, y2, t2) = glider_pair

  dx = x2 - x1
  dy = y2 - y1
  dt = t2 - t1

  hd = dx - dy
  tr = float(hd) / 2 * 47

  return [abs(dt - 4 * dy - tr), abs(4 * dx - dt - tr)]

invariants = get_invariants(g.getcells(g.getselrect()))
g.show(' '.join([str(item) for item in invariants]))
